docs/basics/101-145-hooks.rst

```diff
@@ -0,0 +1,162 @@
+need to be set in ``.datalad/config``:
+
+.. code-block:: bash
+
+   datalad.result-hook.<name>.match-json
+
+and
+
+.. code-block:: bash
+
+   datalad.result-hook.<name>.call-json
+
+Here is what you need to know about these variables:
+
+- The ``<name>`` part of the configurations is the same for both variables, and can be
+  an arbitrarily [#f1]_ chosen name that serves as an identifier for the hook you are
+  defining.
+
+- The first configuration variable, ``datalad.result-hook.<name>.match-json``, defines
+  the requirements that a result evaluation needs to match in order to trigger the hook.
+
+- The second configuration variable, ``datalad.result-hook.<name>.call-json``, defines
+  what the hook execution comprises. It can be any DataLad command of your choice.
+
+And here is how to set the values for these variables:
+
+- The value for ``datalad.result-hook.<name>.match-json`` needs to be specified as
+  a JSON-encoded dictionary with any number of keys, such as
+
+  .. code-block:: bash
+
+     {"type": "file", "action": "get", "status": "notneeded"}
+
+  This translates to: "Match a "not-needed" after :command:`datalad get` of a file."
+  If all specified values in the keys in this dictionary match the values of the
+  same keys in the result evaluation, the hook is executed. Apart from ``==``
+  evaluations, ``in``, ``not in``, and ``!=`` are supported. To make use of such
+  operations, the test value needs to be wrapped into a list, with the first item
+  being the operation, and the second value the test value, such as
+
+  .. code-block:: bash
+
+     {"type": ["in", ["file", "directory"]], "action": "get", "status": "notneeded"}
+
+  This translates to:  "Match a "not-needed" after :command:`datalad get` of a file or directory."
+  Another example is this::
+
+     '{"type":"dataset","action":"install","status":["eq", "ok"]}'
+
+  which translates to: "Match a successful installation of a dataset".
+
+- The value for ``datalad.result-hook.<name>.call-json`` is specified in its
+  Python notation, and its options are specified as a JSON-encoded dictionary
+  with keyword arguments. Conveniently, a number of string substitutions are
+  supported: a ``dsarg`` argument expands to the ``dataset`` given to the initial
+  command the hook operates in, and any key from the result evaluation can be
+  expanded to the respective value in the result dictionary. Curly braces need to
+  be escaped by doubling them.
+  This is not the easiest specification there is, but its also not as hard as it
+  may sound. Here is how this could look like for a :command:`datalad unlock`::
+
+     $ unlock {{"dataset": "{dsarg}", "path": "{path}"}}
+
+  This translates to "unlock the path the previous command operated on, in the
+  dataset the previous command operated on". Another example is this run command::
+
+     $ run  {{"cmd": "touch {path}_annoyed", "dataset": "{dsarg}", "explicit": true}}
```

Comment:
I know this was me, but it is a stupid example ;-)

What about replacing the command with `cp ~/templates/standard-readme.txt {path}/README` and sell it as automatically populate a dataset with a default README.

Reply:
(Personally, I found the example hilarious ;-) Will change it nevertheless)

Reply:
I fail to get such a hook to run because of an inconveniently coinciding mismatching formatting options and command requirements: I am trying to execute "``datalad run --output "README" "cp ~/Templates/standard-readme.txt" --explicit``" via a hook after successful installation of a dataset. Here are my ``git config`` calls:

For matching:
```sh
git config --global --add datalad.result-hook.readme.match-json '{"type": "dataset","action":"create","status":"ok"}'
```

Hook definition:

```sh
git config --global --add datalad.result-hook.readme.call-json 'run {{"cmd":"cp ~/Templates/standard-readme.txt {path}/README", "outputs":"["README"]", "dataset":"{path}","explicit":true}}'
```

The important part here is ``"outputs":"["README"]"``. I need to give the output definition as a list with the string. I *thought* I once PR'ed an ``assure_list()``, but could only find [this](https://github.com/datalad/datalad/pull/3761) for ``create``... 

This fails with (Debug output):

```sh
[WARNING] Invalid argument specification for hook readme (after parameter substitutions): {"cmd":"cp ~/Templates/standard-readme.txt /tmp/ads18/README", "outputs":"["README"]", "dataset":"/tmp/ads18","explicit":true} [Expecting ',' delimiter: line 1 column 77 (char 76) [decoder.py:raw_decode:353]], hook will be skipped 
```

Because in ``"outputs":"["README"]"``, only the ``"["`` part is considered. Switching to single quotes (``"outputs":"['README']"``) or leaving quotation marks completely (both lead to no quotes in the config file) leads to ``README`` being split into its components: 

```sh
[DEBUG  ] Resolved dataset for saving: /tmp/ads21 
[DEBUG  ] Determined class of decorated function: <class 'datalad.core.local.status.Status'> 
[DEBUG  ] Resolved dataset for status reporting: /tmp/ads21 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/A)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/D)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/E)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/E)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/M)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/R)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/[)] 
[DEBUG  ] Resolved dataset for path resolution: /tmp/ads21 
[ERROR  ] path not underneath this dataset [status(/tmp/])] 
[DEBUG  ] Determined 0 datasets for saving from input arguments 
[DEBUG  ] chdir '/tmp' -> '/tmp' (coming back) 
[DEBUG  ] could not perform all requested actions: Command did not complete successfully [{'action': 'status', 'path': PosixPath('/tmp/A'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/D'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/E'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/E'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/M'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/R'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/['), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}, {'action': 'status', 'path': PosixPath('/tmp/]'), 'refds': '/tmp/ads21', 'status': 'error', 'message': 'path not underneath this dataset'}] [utils.py:generator_func:495] 
```

I've unsucessfully tried escaping single or double quotes in the config call as well (e.g., ``git config --global --replace-all datalad.result-hook.readme.call-json 'run {{"cmd":"cp ~/Templates/standard-readme.txt {path}/README", "outputs":"[\'README\']", "dataset":"{path}","explicit":true}}'``. Do you see a way of giving a list with a string to the hook definition @mih?

Reply:
I will try it myself, but why `"["README"]"` instead of `["README"]`. This is valid JSON for a list.

Reply:
Seems to work fine without the outer quotes.

```
% git config --add datalad.result-hook.readme.match-json '{"type": "dataset","action":"create","status":"ok"}'
% git config --add datalad.result-hook.readme.call-json 'run {{"cmd":"cp ~/Templates/standard-readme.txt {path}/README", "outputs":["README"], "dataset":"{path}","explicit":true}}'
% datalad create -d . subds1
[INFO   ] Creating a new annex repo at /tmp/testhook/subds1 
add(ok): subds1 (file)                                                                                                       
add(ok): .gitmodules (file)
save(ok): . (dataset)
create(ok): subds1 (dataset)
[INFO   ] == Command start (output follows) ===== 
[INFO   ] == Command exit (modification check follows) ===== 
(datalad3-dev) 1 mih@meiner /tmp/testhook (git)-[master] % ls subds1
README
(datalad3-dev) mih@meiner /tmp/testhook (git)-[master] % cat subds1/README
dummy

```